upgradability/forward-call/address_book_aggregator/lib.rs

```diff
@@ -0,0 +1,96 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+mod address_book_aggregator {
+    use ink_env::call::{Call, ExecutionInput, Selector};
+    use ink_prelude::string::String;
+
+    type SelectorData = [u8; 4];
+    type ExternalMessageData = Option<(AccountId, SelectorData)>;
+
+    const MAX_BOOK_COUNT: usize = 5;
+    const MAX_RETURNED_INFO_SIZE: usize = 20;
+
+    #[ink(storage)]
+    pub struct AddressBookAggregator {
+        /// Owner of the contract, can change contracts we forward to.
+        owner: AccountId,
+
+        /// Ids of the contract we query, along with required selectors.
+        address_books: [ExternalMessageData; MAX_BOOK_COUNT],
+    }
+
+    impl AddressBookAggregator {
+        /// Inintializes the contract and sets it's owner.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {
+                owner: Self::env().caller(),
+                address_books: [None; MAX_BOOK_COUNT],
+            }
+        }
+
+        /// Allows owner of this contract to add/modify one of
+        /// the addresses that we forward to.
+        /// Returns true if this operation was successful.
+        #[ink(message)]
+        pub fn set_address_book(&mut self, contract_id: AccountId, get_selector: SelectorData, book_id: u64) -> bool {
+            if self.env().caller() != self.owner || book_id as usize > MAX_BOOK_COUNT {
+                return false;
+            }
+
+            self.address_books[book_id as usize] = Some((contract_id, get_selector));
+            true
+        }
+
+        /// Allows owner of this contract to remove one of
+        /// the addresses we forward to.
+        /// Returns true if this operation was successful.
+        #[ink(message)]
+        pub fn remove_address_book(&mut self, book_id: u64) -> bool {
+            if self.env().caller() != self.owner || book_id as usize > MAX_BOOK_COUNT {
+                return false;
+            }
+
+            self.address_books[book_id as usize] = None;
+            true
+        }
+
+        /// A function which queries memorized contracts
+        /// for contact info of a specified address.
+        #[ink(message)]
+        pub fn get_info(&self, account_id: AccountId) -> Option<String> {
+            for id in (0..MAX_BOOK_COUNT).rev() {
+                if let Some((forward_to, selector)) = self.address_books[id] {
+
+                    // Here we perform a forward call to a contract that is supposed
+                    // to store contact info for some addresses.
+                    // If we are able to retrieve this info, then we retrurn it.
+                    // Otherwise we will continue to search in the rest of "address books".
+                    let res = ink_env::call::build_call::<ink_env::DefaultEnvironment>()
```

Comment:
personal opinion, but matching on result would increase readability?

Reply:
So just remove auxiliary variable, or create an actual match that does something only on one pattern?

Reply:
> I am aware that this is not a typical forward proxy, what I thought was interesting in upgradability examples is the difference in storage management. But you are probably right - putting it as a forward-call example might be confusing. Would you recommend moving this example to something like `cross_contract_calls/` and preparing a new one for the forward-call?

Sure, that would work

Reply:
I leave it up to you if you want to just resolve this comment, but for example:
```
                match self.address_books[id] {
                    Some((forward_to, selector)) => {
                        match Self::do_get_info(forward_to, account_id, selector) {
                            Ok(info) => return info,
                            Err(_err) => {
                                // ignore
                            }
                        }
                    }
                    None => {
                        // keep looking
                    }
                }
```